Start using RuboCop for linting

ob-stripe opened the pull request:
@brandur-stripe Here is a first pass at enforcing RuboCop linting in stripe-ruby.

For now, I've set it to use all default rules, with the exception of enforcing commas after the last element in multiline arrays/hashes. This is largely a matter of personal preference so let me know if you'd rather not have this rule, and/or if you want to change other default rules too.

Once we've settled on a set of rules, we can start fixing all the linting errors. Right now, there are 3412 offenses, but a lot of them can be easily auto-corrected.

ob-stripe commented on .gitignore:
```diff
@@ -3,3 +3,4 @@
 .rvmrc
 Gemfile.lock
 tags
+/.bundle/
```
Not _directly_ related to the PR, but after adding `rubocop` to the Gemfile and running `bundle install`, I noticed that the `.bundle` directory wasn't ignored by git.

brandur-stripe replied:
Good call.

brandur-stripe commented:
Hey OB, this is really exciting stuff!

> For now, I've set it to use all default rules, with the exception of enforcing commas after the last element in multiline arrays/hashes. This is largely a matter of personal preference so let me know if you'd rather not have this rule, and/or if you want to change other default rules too.

Totally behind commas after the last element! +1.

> Once we've settled on a set of rules, we can start fixing all the linting errors. Right now, there are 3412 offenses, but a lot of them can be easily auto-corrected.

Nice. I wonder if a good start might be to just try autocorrecting 3-10 meaty files and then committing them here. If any stylistic stuff comes up that we hate, we can talk about it then.

ob-stripe commented on lib/stripe.rb:
```diff
@@ -220,10 +220,8 @@ def self.uri_encode(params)
   class << self
     extend Gem::Deprecate
-    deprecate :uri_encode, "Stripe::Util#encode_parameters", 2016, 01
+    deprecate :uri_encode, "Stripe::Util#encode_parameters", 2016, 0o1
```
Ah-ha! I didn't know that Ruby used `0` as the octal prefix. That clearly wasn't the intention here so we'll need to track those instances and remove the prefixes.

brandur-stripe replied:
Wow, Rubocop's off to a great start with finding problems for us!

ob-stripe commented on lib/stripe/stripe_object.rb:
```diff
@@ -428,15 +420,14 @@ def dirty_value!(value)
-      when StripeObject then obj.instance_variable_get(:@values)
-      else
-        raise ArgumentError, "#empty_values got unexpected object type: #{obj.class.name}"
+               when Hash         then obj
```
Not a big fan of this indent style.

brandur-stripe replied:
Yeah I don't like this, the trouble is at any point you rename `values` then every single line has to change.

ob-stripe replied:
Switched `Layout/CaseIndentation` to `end` to solve this.

ob-stripe commented on lib/stripe/util.rb:
```diff
@@ -91,47 +91,50 @@ def self.convert_to_stripe_object(data, opts = {})
+         !Stripe.log_level.nil? && Stripe.log_level <= Stripe::LEVEL_ERROR
         log_internal(message, data, color: :cyan,
-          level: Stripe::LEVEL_ERROR, logger: Stripe.logger, out: $stderr)
+                                    level: Stripe::LEVEL_ERROR, logger: Stripe.logger, out: $stderr)
```
Also not a huge fan of this indent style. I'd prefer using either a 2-space indent or align the second line with the first parameter, not with the first _keyword_ parameter.

brandur-stripe replied:
+1. For the same reason as above, it's just impractical in that it doesn't do anything to improve readability (over two-space style), but creates unnecessarily large diffs anytime the method or any of these variables changes name.

ob-stripe replied:
Okay, so the problem with this one is that named parameters implicitly form a hash, so they're handled by hash alignment rules. AFAICT, it's possible to tell Rubocop to ignore or inspect hashes as arguments (implicit or explicit) but not to apply different alignment rules than for "normal" hashes.

A simple fix is to start named parameters on a newline, but it's a shame that Rubocop can't handle that automatically for us.

brandur-stripe replied:
Ah, cool. Thanks for investigating!

Honestly in this case, `color` would probably be better off started on a newline anyway.

ob-stripe commented on lib/stripe/util.rb:
```diff
@@ -91,47 +91,50 @@ def self.convert_to_stripe_object(data, opts = {})
+    rescue
+      false
+    else
+      true
```
I don't really like the `def/rescue/end` pattern.

brandur-stripe replied:
I can see the argument that it obscures the presence of your `rescue` block, but IMO these aren't too bad as long as methods are kept relatively short.

ob-stripe commented on lib/stripe/util.rb:
```diff
@@ -91,47 +91,50 @@ def self.convert_to_stripe_object(data, opts = {})
+                   key.to_sym
+                 rescue
+                   key
+                 end) || key
```
I don't like what autocorrect did here either -- the one-liner was pretty straightforward and legible.

brandur-stripe replied:
Yeah this is crazy haha. Curious to see what it does after we remove the "aligned indent" rule or whatever it is.

ob-stripe replied:
I can't find the rule that caused this, but maybe using `respond_to?` would be better practice anyway?

brandur-stripe replied:
> I can't find the rule that caused this, but maybe using respond_to? would be better practice anyway?

Yeah, much better. Also, using an inline `rescue` like that was a little sketchy to start. Should be fine.

ob-stripe commented:
@brandur-stripe Okay, I pushed another commit where I ran the autocorrect on 4 different files and left a few comments where I didn't like the corrections. Looking forward to hearing your opinion :)

brandur-stripe commented on lib/stripe.rb:
```diff
@@ -220,10 +220,8 @@ def self.uri_encode(params)
-unless ENV["STRIPE_LOG"].nil?
-  Stripe.log_level = ENV["STRIPE_LOG"]
-end
+Stripe.log_level = ENV["STRIPE_LOG"] unless ENV["STRIPE_LOG"].nil?
```
I guess pulling things onto one line is alright, but seems kind of unnecessary. Probably not worth changing though.

brandur-stripe commented on lib/stripe/account.rb:
```diff
@@ -70,7 +70,7 @@ def serialize_params(options = {})
     end
 
-    def serialize_params_account(obj, update_hash)
+    def serialize_params_account(_obj, update_hash)
```
Love the unused variable convention.

brandur-stripe commented on lib/stripe/util.rb:
```diff
@@ -147,8 +150,8 @@ def self.symbolize_names(object)
-      Util.flatten_params(params).
-        map { |k,v| "#{url_encode(k)}=#{url_encode(v)}" }.join('&')
+      Util.flatten_params(params)
+          .map { |k, v| "#{url_encode(k)}=#{url_encode(v)}" }.join('&')
```
This prefixed dot didn't use to be possible on previous versions of Ruby which has led me to still not really use it, but that said, I kind of like the change.

brandur-stripe commented:
> @brandur-stripe Okay, I pushed another commit where I ran the autocorrect on 4 different files and left a few comments where I didn't like the corrections. Looking forward to hearing your opinion :)

Thanks :) I agree with you in that I dislike the "aligned indent to first parameter" style, but otherwise I think the changes are pretty positive.

brandur-stripe commented:
@ob-stripe Should we try to pull in at least a preliminary version of this one?

ob-stripe commented:
Ah, sorry, I meant to get back to this. I've identified some of the rules we need to change to get closer to the style we want to enforce. I'll try to push an update tomorrow.

ob-stripe commented:
@brandur-stripe Okay, I added a rule for aligning `case` and re-run `--auto-correct` and `--auto-gen-config`. Can you tell me what you think?

If we're ~happy with the current state, I think we can pull this in to start enforcing the rules for new code, and start progressively solving all the issues listed in `.rubocop_todo.yml`.

brandur-stripe commented:
> If we're ~happy with the current state, I think we can pull this in to start enforcing the rules for new code, and start progressively solving all the issues listed in .rubocop_todo.yml.

@ob-stripe LGTM. If you're happy, I'd say let's go for the merge and we'll take things from there.

ob-stripe commented:
@brandur-stripe Okay, I've added a couple of rules to get closer to what I think we want, applied auto-correct to all files and removed the WIP tag.

Obviously the PR is pretty painful to review :( If you prefer, we could do this in stages and create a few distinct PRs to apply some of the more impactful (in terms of numbers of changed lines) one by one, e.g. using the newer hash syntax or using single quotes for strings.

brandur-stripe commented on lib/stripe/util.rb:
```diff
@@ -285,16 +287,16 @@ def self.secure_compare(a, b)
+      cyan: 6, light_cyan: 66,
+      white: 7, light_white: 67,
+      default: 9,
+    }.freeze
```
We're switching to colon syntax, but would you mind checking if we can keep the alignment on this hash? I think it's quite a bit easier to read before to the cop fix.

ob-stripe replied:
I don't think there are any rules specifically for multi-line hashes with more than one entry per line. I manually re-aligned the hash and Rubocop is still happy with it, so ¯\\_(ツ)_/¯

brandur-stripe commented:
Thanks OB!

Left one minor comment above, but mostly looks good to me. Can we also add: (1) a `.travis.yml` entry so that checks will run in CI and (2) an entry in the README explaining how to run Rubocop locally?

Aside from that, the one other thing I noticed was all the conversions to single quoted strings. I guess I don't feel too strongly about this, but I don't think the single quoted strings at a whole of value (the performance improvement is negligible) and they can be annoying at times if you forget to change them when you're adding an interpolation. Not sure how you feel on that front.

ob-stripe commented:
I fixed the hash alignment you pointed out, and added a rule for enforcing double quotes instead of single quotes for strings and re-run auto-correct afterwards. I don't feel too strongly about this personally -- I tend to use single quotes for "code" strings and double quotes for strings that are meant for human consumption, but that's just a personal habit I carried over from Python.

If you prefer, we can disable this rule and let developers use single or double quotes as they see fit.

I added a note to the README to show how to run the linter. We don't need to modify `.travis.yml` -- it already runs `bundle exec rake` which executes all Rake tasks, including the Rubocop one I added in the Rakefile.

brandur-stripe commented:
> I fixed the hash alignment you pointed out, and added a rule for enforcing double quotes instead of single quotes for strings and re-run auto-correct afterwards.

Cool! Actually, I think that personally I like this the most.

> If you prefer, we can disable this rule and let developers use single or double quotes as they see fit.

Nope, this solution sounds good.

> I added a note to the README to show how to run the linter. We don't need to modify .travis.yml -- it already runs bundle exec rake which executes all Rake tasks, including the Rubocop one I added in the Rakefile.

Thanks! And great — I'd seen the `Rakefile` addition, but then totally forgot about it. Looks good.
